### PR TITLE
#50.4 schedule

### PR DESCRIPTION
**Pull request by Evgentik**

### CODE REVIEW

File: Recrutify/Recrutify.DataAccess/Models/Interview.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace Recrutify.DataAccess.Models
+{
+    public class Interview
+    {
+        public Guid UserId { get; set; }
+
+        public Guid CandidateId { get; set; }
+
+        public DateTime AppointDateTimeUtc { get; set; }
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.DataAccess/Models/Interview.cs:15]: AppoitmentDateTime

File: Recrutify/Recrutify.DataAccess/Repositories/ScheduleRepository.cs [hunk 2]
```diff
@@ -56,6 +57,27 @@ public Task BulkUpdateScheduleSlotsAsync(Guid userId, IEnumerable<DateTime> newL
             return GetCollection().BulkWriteAsync(updateModelsWithNewScheduleSlot.Union(updateModelsWithRemovedUsers));
         }
 
+        public Task BulkСancelInterviewsAsync(IEnumerable<Interview> cancelledInterviews)
+        {
+            ScheduleCandidateInfo scheduleCandidateInfo = null;
+            var updateBuilder = Builders<Schedule>.Update;
+            var updateModels = cancelledInterviews.Select(i => new UpdateOneModel<Schedule>(
+                _filterBuilder.Eq(s => s.Id, i.UserId),
+                updateBuilder.Set("ScheduleSlots.$[scheduleSlots].ScheduleCandidateInfo", scheduleCandidateInfo))
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.DataAccess/Repositories/ScheduleRepository.cs:66]: just write null, not necessary to use scheduleCandidateInfo

File: Recrutify/Recrutify.DataAccess/Repositories/ScheduleRepository.cs [hunk 3]
```diff
@@ -56,6 +57,27 @@ public Task BulkUpdateScheduleSlotsAsync(Guid userId, IEnumerable<DateTime> newL
...
+            { ArrayFilters = new List<ArrayFilterDefinition>() { new BsonDocumentArrayFilterDefinition<ScheduleSlot>(new BsonDocument("scheduleSlots.AvailableTime", i.AppointDateTimeUtc)) } });
+            return GetCollection().BulkWriteAsync(updateModels);
+        }
+
+        public Task BulkAppointInterviewsAsync(IEnumerable<Interview> cancelledInterviews, IEnumerable<ScheduleCandidateInfo> candidatesInfo)
+        {
+            var updateBuilder = Builders<Schedule>.Update;
+            var updateModels = cancelledInterviews.Select(i => new UpdateOneModel<Schedule>(
+                _filterBuilder.Eq(s => s.Id, i.UserId),
+                updateBuilder.Set("ScheduleSlots.$[scheduleSlots].ScheduleCandidateInfo", candidatesInfo.FirstOrDefault(c => c.Id == i.CandidateId)))
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.DataAccess/Repositories/ScheduleRepository.cs:76]: scheduleSlot wverywhere

File: Recrutify/Recrutify.Host/Controllers/SchedulesController.cs [hunk 4]
```diff
@@ -53,5 +54,21 @@ public async Task<ActionResult> UpdateScheduleSlotsAsync(IEnumerable<DateTime> d
...
+
+        [Authorize(Policy = Constants.Policies.FeedbackPolicy)]
+        [HttpPut("bulk/update_assigned_interviews")]
+        public async Task<ActionResult> BulkAppointOrCancelInterviewsAsync([FromBody] IEnumerable<InterviewDTO> interviews, [FromQuery, Required] Guid projectId, CancellationToken cancellationToken)
+        {
+            try
+            {
+                await _scheduleService.ProcessingAppointOrCancelInterviewsAsync(interviews, projectId, cancellationToken);
+            }
+            catch (Exception e)
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Host/Controllers/SchedulesController.cs:66]: ValidationException

File: Recrutify/Recrutify.Services/DTOs/InterviewDTO.cs [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace Recrutify.Services.DTOs
+{
+    public class InterviewDTO
+    {
+        public Guid UserId { get; set; }
+
+        public Guid CandidateId { get; set; }
+
+        public DateTime AppointDateTimeUtc { get; set; }
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/DTOs/InterviewDTO.cs:15]: AppoitmentDateTime

File: Recrutify/Recrutify.Services/Events/StatusChangeEventProcessor.cs [hunk 6]
```diff
@@ -26,7 +26,7 @@ public void Subscribe()
 
         public async Task UpdateCandidatesStatusesAsync(UpdateStatusEventArgs e)
         {
-            var candidates = await _candidateService.GetCandidatesByIdsAsync(e.CandidatesIds);
+            var candidates = await _candidateService.GetCandidatesDTOByIdsAsync(e.CandidatesIds);
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Events/StatusChangeEventProcessor.cs:29]: just GetCandidatesByIdsAsync will be enougth

File: Recrutify/Recrutify.Services/Services/CandidateService.cs [hunk 7]
```diff
@@ -229,5 +229,15 @@ public async Task<CandidatesProjectInfoDTO> GetCandidatesProjectInfoAsync(Guid?
            candidatesProjectInfo.ProjectName = projectName;
            return candidatesProjectInfo;
         }
+
+        public async Task BulkUpdateStatusAsync(IEnumerable<Guid> candidatesIds, Guid projectId, Status status)
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/CandidateService.cs:233]: async await

File: Recrutify/Recrutify.Services/Services/CandidateService.cs [hunk 8]
```diff
@@ -229,5 +229,15 @@ public async Task<CandidatesProjectInfoDTO> GetCandidatesProjectInfoAsync(Guid?
...
+        {
+            await _candidateRepository.BulkUpdateStatusAsync(candidatesIds, projectId, status);
+        }
+
+        public async Task<IEnumerable<Candidate>> GetCandidatesByIdsAsync(IEnumerable<Guid> ids)
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/CandidateService.cs:238]: async await

File: Recrutify/Recrutify.Services/Services/UserService.cs [hunk 9]
```diff
@@ -63,5 +63,10 @@ public Task BulkUpdateProjectRolesAsync(Guid projectId, IDictionary<Guid, IEnume
             var updateUsers = _staffHelper.GetUpdatedUsersRoles(currentUsersRoles, newUsersRoles);
             return _userRepository.BulkUpdateProjectRolesAsync(projectId, newUsers, remoteUsers, updateUsers);
         }
+
+        public async Task<IEnumerable<UserShort>> GetUsersShortByIdsAsync(IEnumerable<Guid> ids)
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/UserService.cs:67]: async await

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs [hunk 10]
```diff
@@ -81,6 +93,81 @@ public async Task UpdateScheduleSlotsForCurrentUserAsync(IEnumerable<DateTime> d
             await _scheduleRepository.BulkUpdateScheduleSlotsAsync(currentUserId, newListDateTime, removedListDateTime);
         }
 
+        public async Task ProcessingAppointOrCancelInterviewsAsync(IEnumerable<InterviewDTO> interviews, Guid projectId, CancellationToken cancellationToken)
+        {
+            var projectExists = await _projectService.ExistsAsync(projectId, cancellationToken);
+            if (!projectExists)
+            {
+                throw new Exception("Project does not exist.");
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/ScheduleService.cs:101]: ValidationException

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs [hunk 11]
```diff
@@ -81,6 +93,81 @@ public async Task UpdateScheduleSlotsForCurrentUserAsync(IEnumerable<DateTime> d
...
+            }
+
+            var allCadidates = await _candidateService.GetCandidatesByIdsAsync(interviews.Select(i => i.CandidateId));
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/ScheduleService.cs:104]: allCandidates

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs [hunk 12]
```diff
@@ -81,6 +93,81 @@ public async Task UpdateScheduleSlotsForCurrentUserAsync(IEnumerable<DateTime> d
...
+
+            var cancelledInterviews = _mapper.Map<IEnumerable<Interview>>(interviews.Where(i => i.IsAppointment == false));
+            var cadidatesIdsCancelledInterviews = cancelledInterviews.Select(i => i.CandidateId);
+            var cuurentStatusCadidates = allCadidates.Where(c => cadidatesIdsCancelledInterviews.Contains(c.Id)).Select(c => c.ProjectResults.FirstOrDefault(p => p.ProjectId == projectId).Status).FirstOrDefault();
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/ScheduleService.cs:108]: no sense to select all, and only at the end do first or default, just replace where with first or default

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs [hunk 13]
```diff
@@ -81,6 +93,81 @@ public async Task UpdateScheduleSlotsForCurrentUserAsync(IEnumerable<DateTime> d
...
+            var projectExists = await _projectService.ExistsAsync(projectId, cancellationToken);
+            if (!projectExists)
+            {
+                throw new Exception("Project does not exist.");
+            }
+
+            var allCadidates = await _candidateService.GetCandidatesByIdsAsync(interviews.Select(i => i.CandidateId));
+
+            var cancelledInterviews = _mapper.Map<IEnumerable<Interview>>(interviews.Where(i => i.IsAppointment == false));
+            var cadidatesIdsCancelledInterviews = cancelledInterviews.Select(i => i.CandidateId);
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/ScheduleService.cs:107]: there is sense to move 107-108 to if on 110

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs [hunk 14]
```diff
@@ -81,6 +93,81 @@ public async Task UpdateScheduleSlotsForCurrentUserAsync(IEnumerable<DateTime> d
...
+            var cuurentStatusCadidates = allCadidates.Where(c => cadidatesIdsCancelledInterviews.Contains(c.Id)).Select(c => c.ProjectResults.FirstOrDefault(p => p.ProjectId == projectId).Status).FirstOrDefault();
+
+            if (cancelledInterviews.Any())
+            {
+                await _scheduleRepository.BulkСancelInterviewsAsync(cancelledInterviews);
+                await _candidateService.BulkUpdateStatusAsync(cancelledInterviews.Select(i => i.CandidateId), projectId, _statusHelper.GetStatusDown(cuurentStatusCadidates));
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/ScheduleService.cs:113]: in line 107 you've already done select ids

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs [hunk 15]
```diff
@@ -81,6 +93,81 @@ public async Task UpdateScheduleSlotsForCurrentUserAsync(IEnumerable<DateTime> d
...
+        {
+            var projectExists = await _projectService.ExistsAsync(projectId, cancellationToken);
+            if (!projectExists)
+            {
+                throw new Exception("Project does not exist.");
+            }
+
+            var allCadidates = await _candidateService.GetCandidatesByIdsAsync(interviews.Select(i => i.CandidateId));
+
+            var cancelledInterviews = _mapper.Map<IEnumerable<Interview>>(interviews.Where(i => i.IsAppointment == false));
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/ScheduleService.cs:106]: try to create dictionary with the key of bool type, then it wouldn't be necessary to do Where twice for true and false

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs [hunk 16]
```diff
@@ -81,6 +93,81 @@ public async Task UpdateScheduleSlotsForCurrentUserAsync(IEnumerable<DateTime> d
...
+            if (cancelledInterviews.Any())
+            {
+                await _scheduleRepository.BulkСancelInterviewsAsync(cancelledInterviews);
+                await _candidateService.BulkUpdateStatusAsync(cancelledInterviews.Select(i => i.CandidateId), projectId, _statusHelper.GetStatusDown(cuurentStatusCadidates));
+            }
+
+            var appointedInterviews = _mapper.Map<IEnumerable<Interview>>(interviews.Where(i => i.IsAppointment == true));
+            var users = await _userService.GetUsersShortByIdsAsync(appointedInterviews.Select(i => i.UserId));
+            var cadidatesIdsAppointedInterviews = appointedInterviews.Select(i => i.CandidateId);
+            cuurentStatusCadidates = allCadidates.Where(c => cadidatesIdsAppointedInterviews.Contains(c.Id)).Select(c => c.ProjectResults.FirstOrDefault(p => p.ProjectId == projectId).Status).FirstOrDefault();
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/ScheduleService.cs:119]: same as 108

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs [hunk 17]
```diff
@@ -81,6 +93,81 @@ public async Task UpdateScheduleSlotsForCurrentUserAsync(IEnumerable<DateTime> d
...
+
+            var usersForInvite = users.Select(c => new UserEmailInfo() { Id = c.Id, Email = c.Email, Name = c.Name }).ToList();
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/ScheduleService.cs:121]: you could use here UserShort, why you need to recopy it to  UserEmailInfo

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs [hunk 18]
```diff
@@ -81,6 +93,81 @@ public async Task UpdateScheduleSlotsForCurrentUserAsync(IEnumerable<DateTime> d
...
+            var candidatesForInvite = allCadidates.Where(c => cadidatesIdsAppointedInterviews.Contains(c.Id))
+                                        .Select(c => new CandidateEmailInfo()
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/ScheduleService.cs:123]: use mapping

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs [hunk 19]
```diff
@@ -81,6 +93,81 @@ public async Task UpdateScheduleSlotsForCurrentUserAsync(IEnumerable<DateTime> d
...
+                                    Candidate = candidatesForInvite.FirstOrDefault(c => c.Id == i.CandidateId),
+                                    User = usersForInvite.FirstOrDefault(u => u.Id == i.UserId),
+                                }),
+            };
+
+            if (appointedInterviews.Any())
+            {
+                await _scheduleRepository.BulkAppointInterviewsAsync(
+                    appointedInterviews,
+                    allCadidates.Where(c => cadidatesIdsAppointedInterviews.Contains(c.Id)).Select(c => new ScheduleCandidateInfo()
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/ScheduleService.cs:148]: use mapping

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs [hunk 20]
```diff
@@ -81,6 +93,81 @@ public async Task UpdateScheduleSlotsForCurrentUserAsync(IEnumerable<DateTime> d
...
+                    {
+                         BestTimeToConnect = c.BestTimeToConnect,
+                         Id = c.Id,
+                         Name = c.Name,
+                         Surname = c.Surname,
+                         Email = c.Email,
+                         Skype = c?.Contacts?.FirstOrDefault(s => s.Type == "Skype")?.Value,
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/ScheduleService.cs:155]: why you write c?, if in prev lines you already wrote c.Email for example, if it woulf null exception, ? won't save you on last line

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs [hunk 21]
```diff
@@ -81,6 +93,81 @@ public async Task UpdateScheduleSlotsForCurrentUserAsync(IEnumerable<DateTime> d
...
+                         ProjectResult = c.ProjectResults
+                                              .Select(p => new ScheduleCandidateProjectResult()
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/ScheduleService.cs:157]: mapping

File: Recrutify/Recrutify.DataAccess/Models/Interview.cs (hunk 1 above)
Comment (Evgentik) [reply Recrutify/Recrutify.DataAccess/Models/Interview.cs:15]: +

File: Recrutify/Recrutify.DataAccess/Repositories/ScheduleRepository.cs (hunk 2 above)
Comment (Evgentik) [reply Recrutify/Recrutify.DataAccess/Repositories/ScheduleRepository.cs:66]: +

File: Recrutify/Recrutify.DataAccess/Repositories/ScheduleRepository.cs (hunk 3 above)
Comment (Evgentik) [reply Recrutify/Recrutify.DataAccess/Repositories/ScheduleRepository.cs:76]: +

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs (hunk 11 above)
Comment (Evgentik) [reply Recrutify/Recrutify.Services/Services/ScheduleService.cs:104]: +

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs (hunk 13 above)
Comment (Evgentik) [reply Recrutify/Recrutify.Services/Services/ScheduleService.cs:107]: +

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs (hunk 20 above)
Comment (Evgentik) [reply Recrutify/Recrutify.Services/Services/ScheduleService.cs:155]: +

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs (hunk 18 above)
Comment (Evgentik) [reply Recrutify/Recrutify.Services/Services/ScheduleService.cs:123]: +

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs (hunk 17 above)
Comment (Evgentik) [reply Recrutify/Recrutify.Services/Services/ScheduleService.cs:121]: +

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs (hunk 16 above)
Comment (Evgentik) [reply Recrutify/Recrutify.Services/Services/ScheduleService.cs:119]: +

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs (hunk 15 above)
Comment (Evgentik) [reply Recrutify/Recrutify.Services/Services/ScheduleService.cs:106]: +

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs (hunk 14 above)
Comment (Evgentik) [reply Recrutify/Recrutify.Services/Services/ScheduleService.cs:113]: +

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs (hunk 12 above)
Comment (Evgentik) [reply Recrutify/Recrutify.Services/Services/ScheduleService.cs:108]: +

File: Recrutify/Recrutify.Services/Events/StatusChangeEventProcessor.cs (hunk 6 above)
Comment (Evgentik) [reply Recrutify/Recrutify.Services/Events/StatusChangeEventProcessor.cs:29]: +

File: Recrutify/Recrutify.Host/Controllers/SchedulesController.cs (hunk 4 above)
Comment (Evgentik) [reply Recrutify/Recrutify.Host/Controllers/SchedulesController.cs:66]: +

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs (hunk 10 above)
Comment (Evgentik) [reply Recrutify/Recrutify.Services/Services/ScheduleService.cs:101]: +

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs (hunk 19 above)
Comment (Evgentik) [reply Recrutify/Recrutify.Services/Services/ScheduleService.cs:148]: +

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs (hunk 21 above)
Comment (Evgentik) [reply Recrutify/Recrutify.Services/Services/ScheduleService.cs:157]: +

File: Recrutify/Recrutify.Services/Services/CandidateService.cs (hunk 8 above)
Comment (Evgentik) [reply Recrutify/Recrutify.Services/Services/CandidateService.cs:238]: +

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs [hunk 22]
```diff
@@ -80,9 +92,80 @@ public async Task UpdateScheduleSlotsForCurrentUserAsync(IEnumerable<DateTime> d
...
+            var projectExists = await _projectService.ExistsAsync(projectId, cancellationToken);
+            if (!projectExists)
+            {
+                throw new ValidationException("Project does not exist.");
+            }
+
+            var allCandidates = await _candidateService.GetCandidatesByIdsAsync(interviews.Select(i => i.CandidateId));
+            var dictionaryInterviews = _mapper.Map<IEnumerable<Interview>>(interviews)
+                                                .GroupBy(i => i.IsAppointment)
+                                                .ToDictionary(k => k.Key, v => v.Select(i => i).ToList());
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/ScheduleService.cs:106]: write just v.ToList(), select not necessary

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs [hunk 23]
```diff
@@ -80,9 +92,80 @@ public async Task UpdateScheduleSlotsForCurrentUserAsync(IEnumerable<DateTime> d
...
+
+            var canceledInterviews = dictionaryInterviews.TryGetValue(false, out var i) ? i : default;
+            var appointedtInterviews = dictionaryInterviews.TryGetValue(true, out var l) ? l : default;
+
+            if (canceledInterviews != null)
+            {
+                await BulkСancelInterviewsAsync(projectId, allCandidates.Where(c => canceledInterviews.Select(i => i.CandidateId).Contains(c.Id)), canceledInterviews);
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/ScheduleService.cs:113]: Move this select to separate var. remember what on each iteration of where you do select again and again

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs [hunk 24]
```diff
@@ -80,9 +92,80 @@ public async Task UpdateScheduleSlotsForCurrentUserAsync(IEnumerable<DateTime> d
...
+            {
+                await BulkAppointInterviewsAsync(projectId, allCandidates.Where(c => appointedtInterviews.Select(i => i.CandidateId).Contains(c.Id)), appointedtInterviews);
+            }
         }
+
+        private async Task BulkСancelInterviewsAsync(Guid projectId, IEnumerable<Candidate> candidates, IEnumerable<Interview> interviews)
+        {
+            var cuurentStatusCadidates = candidates.FirstOrDefault().ProjectResults.FirstOrDefault(p => p.ProjectId == projectId).Status;
+
+            if (!interviews.Any())
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/ScheduleService.cs:126]: Move to the beginning of method, there is a case when you do prev line without sense

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs [hunk 25]
```diff
@@ -80,9 +92,80 @@ public async Task UpdateScheduleSlotsForCurrentUserAsync(IEnumerable<DateTime> d
...
+            {
+                return;
+            }
+
+            await _scheduleRepository.BulkСancelInterviewsAsync(interviews);
+            await _candidateService.BulkUpdateStatusAsync(interviews.Select(i => i.CandidateId), projectId, _statusHelper.GetStatusDown(cuurentStatusCadidates));
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/ScheduleService.cs:132]: You calculate Ids again, you already did it, just pass to this Method Ids

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs [hunk 26]
```diff
@@ -80,9 +92,80 @@ public async Task UpdateScheduleSlotsForCurrentUserAsync(IEnumerable<DateTime> d
...
+        }
+
+        private async Task BulkAppointInterviewsAsync(Guid projectId, IEnumerable<Candidate> candidates, IEnumerable<Interview> interviews)
+        {
+            var cuurentStatusCandidates = candidates.FirstOrDefault().ProjectResults.FirstOrDefault(p => p.ProjectId == projectId).Status;
+            var usersForInvite = await _userService.GetUsersShortByIdsAsync(interviews.Select(i => i.UserId));
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/ScheduleService.cs:138]: Select was already calculated

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs [hunk 27]
```diff
@@ -80,9 +92,80 @@ public async Task UpdateScheduleSlotsForCurrentUserAsync(IEnumerable<DateTime> d
...
+                    interviews,
+                    candidates.Select(c =>
+                    {
+                        var candidateInfo = _mapper.Map<ScheduleCandidateInfo>(c);
+                        var projectResult = c.ProjectResults.FirstOrDefault(p => p.ProjectId == projectId);
+                        projectResult.Status = _statusHelper.GetStatusUp(projectResult.Status);
+                        candidateInfo.ProjectResult = _mapper.Map<ScheduleCandidateProjectResult>(projectResult);
+                        return candidateInfo;
+                    }));
+            await _candidateService.BulkUpdateStatusAsync(interviews.Select(i => i.CandidateId), projectId, _statusHelper.GetStatusUp(cuurentStatusCandidates));
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/ScheduleService.cs:162]: Calculated

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs [hunk 28]
```diff
@@ -80,9 +92,80 @@ public async Task UpdateScheduleSlotsForCurrentUserAsync(IEnumerable<DateTime> d
...
+            _inviteEventPublisher.OnAssignedInterview(args);
+        }
+
+        private IEnumerable<DateTime> GetDateTimeInScheduleSlots(IEnumerable<ScheduleSlot> scheduleSlots)
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/ScheduleService.cs:166]: Is it used ?

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs [hunk 29]
```diff
@@ -80,9 +92,80 @@ public async Task UpdateScheduleSlotsForCurrentUserAsync(IEnumerable<DateTime> d
...
+            {
+                throw new ValidationException("Project does not exist.");
+            }
+
+            var allCandidates = await _candidateService.GetCandidatesByIdsAsync(interviews.Select(i => i.CandidateId));
+            var dictionaryInterviews = _mapper.Map<IEnumerable<Interview>>(interviews)
+                                                .GroupBy(i => i.IsAppointment)
+                                                .ToDictionary(k => k.Key, v => v.Select(i => i).ToList());
+
+            var canceledInterviews = dictionaryInterviews.TryGetValue(false, out var i) ? i : default;
```
Comment (elizavetapavliv) [inline Recrutify/Recrutify.Services/Services/ScheduleService.cs:108]: I and l - give meaningful names

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs (hunk 22 above)
Comment (Evgentik) [reply Recrutify/Recrutify.Services/Services/ScheduleService.cs:106]: +

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs (hunk 23 above)
Comment (Evgentik) [reply Recrutify/Recrutify.Services/Services/ScheduleService.cs:113]: +

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs [hunk 30]
```diff
@@ -80,9 +92,80 @@ public async Task UpdateScheduleSlotsForCurrentUserAsync(IEnumerable<DateTime> d
...
+            var appointedtInterviews = dictionaryInterviews.TryGetValue(true, out var l) ? l : default;
+
+            if (canceledInterviews != null)
+            {
+                await BulkСancelInterviewsAsync(projectId, allCandidates.Where(c => canceledInterviews.Select(i => i.CandidateId).Contains(c.Id)), canceledInterviews);
+            }
+
+            if (appointedtInterviews != null)
+            {
+                await BulkAppointInterviewsAsync(projectId, allCandidates.Where(c => appointedtInterviews.Select(i => i.CandidateId).Contains(c.Id)), appointedtInterviews);
```
[the remark this answers is not in the document]
Comment (Evgentik) [reply Recrutify/Recrutify.Services/Services/ScheduleService.cs:118]: +

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs (hunk 24 above)
Comment (Evgentik) [reply Recrutify/Recrutify.Services/Services/ScheduleService.cs:126]: +

File: Recrutify/Recrutify.Services/Services/ScheduleService.cs (hunk 25 above)
Comment (Evgentik) [reply Recrutify/Recrutify.Services/Services/ScheduleService.cs:132]: +

**Review by elizavetapavliv — APPROVED** [review]

**Pull request merged**